translate/storage/php.py

```diff
@@ -216,171 +339,88 @@ def serialize(self, out):
-
-                # If the translation is already parsed, save it and initialize
-                # a new translation unit.
-                if not invalue:
-                    self.addunit(newunit)
-                    value = ""
-                    newunit = phpunit()
-
-                # Update end delimiter position to the previous last appearance
-                # of the end delimiter, because it might be several entries in
-                # the same line.
-                enddelpos = value.rfind(enddel, 0, enddelpos)
-            else:
-                # After processing current line, if we are not in an array,
-                # fall back to default dialect (PHP simple variable syntax).
-                if not inarray:
-                    equaldel = "="
-                    enddel = ";"
-
-            # If this is part of a multiline translation, just append it to the
-            # previous translation lines.
-            if invalue:
-                lastvalue = lastvalue + value + "\n"
+        def newunit(name, value, escape_type, comments):
+            newunit = phpunit()
+            newunit.escape_type = escape_type
+            newunit.addlocation(name)
+            newunit.source = value
+            for comment in comments:
+                newunit.addnote(comment, 'developer')
+            self.addunit(newunit)
+
+        def format_key(key):
+            if isinstance(key, (int, float)):
+                return '{0}'.format(key)
+            return '\'{0}\''.format(key)
+
+        def handle_array(prefix, nodes, lexer):
+            for item in nodes:
+                assert isinstance(item, ArrayElement)
+                # Skip empty keys
+                if item.key == '':
+                    continue
+                name = format_key(item.key)
+                if prefix:
+                    name = '{0}->{1}'.format(prefix, name)
+                if isinstance(item.value, Array):
+                    handle_array(name, item.value.nodes, lexer)
+                elif isinstance(item.value, six.string_types):
+                    newunit(
+                        name,
+                        item.value,
+                        lexer.extract_quote(),
+                        lexer.extract_comments(item.lexpositions[1]),
+                    )
+
+        def concatenate(item):
+            if isinstance(item, six.string_types):
+                return item
+            assert isinstance(item, BinaryOp)
+            return concatenate(item.left) + concatenate(item.right)
+
+        parser = make_parser()
+        for item in parser.productions:
+            item.callable = wrap_production(item.callable)
+        lexer = PHPLexer()
+        pos = 0
+        tree = parser.parse(phpsrc.decode(self.encoding), lexer=lexer, tracking=True)
+        # Handle text without PHP start
+        if len(tree) == 1 and isinstance(tree[0], InlineHTML):
+            return self.parse(b'<?php\n' + phpsrc)
```

Comment:
Is it necessary to call `self.parse`? Doesn't the following work?:

```python
            phpsrc = b'<?php\n' + phpsrc
            tree = parser.parse(phpsrc.decode(self.encoding), lexer=lexer, tracking=True)
```

This way we should be able to avoid initializing stuff again.

Reply:
The lexer has internal state, so it should be reinitialized. I'm not sure about parser, so I've rather did it the safe way...

Reply:
I am ok with leaving it as is.